### PR TITLE
Unbuffered channels

### PR DESCRIPTION
**Pull request by SimonMeldahl**
Fixed issue #722.
Implemented unbuffered channels to allow synchronous communication between processes.

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/runtime/interpreter/Channel.java [hunk 1]
```diff
@@ -48,11 +48,20 @@
    */
   private int bufferSize;
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/runtime/interpreter/Channel.java]: can this be final?

File: main/src/ca/uwaterloo/flix/runtime/interpreter/Channel.java [hunk 2]
```diff
@@ -48,11 +48,20 @@
...
 
+  /**
+   * a flag for whether the channel is unbuffered or not
+   */
+  private boolean unbuffered = false;
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/runtime/interpreter/Channel.java]: can this be final?

File: main/src/ca/uwaterloo/flix/runtime/interpreter/Channel.java [hunk 3]
```diff
@@ -48,11 +48,20 @@
...
+
   public Channel(int bufferSize) {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/runtime/interpreter/Channel.java]: Does the weeder enforce that channels have a positive size? or how does that work?

**Comment by magnus-madsen** [comment]
Can you change the grammar such that if you dont specify a channel size it defaults to zero?

File: main/src/ca/uwaterloo/flix/runtime/interpreter/Channel.java (hunk 3 above)
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/runtime/interpreter/Channel.java]: It isn't enforced until the Channel constructor is called, which throws an exception if size is less than 0.

File: main/src/ca/uwaterloo/flix/runtime/interpreter/Channel.java (hunk 2 above)
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/runtime/interpreter/Channel.java]: We change it in the constructor if we're given a size of 0 so i don't think so?

File: main/src/ca/uwaterloo/flix/runtime/interpreter/Channel.java (hunk 1 above)
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/runtime/interpreter/Channel.java]: We change it in the constructor if we're given a size of 0 where we use size 1 and set the unbuffered flag, so i don't think so?

Comment (SimonMeldahl) [reply main/src/ca/uwaterloo/flix/runtime/interpreter/Channel.java]: We've made a new commit changing this to final

File: main/src/ca/uwaterloo/flix/runtime/interpreter/Channel.java (hunk 2 above)
Comment (SimonMeldahl) [reply main/src/ca/uwaterloo/flix/runtime/interpreter/Channel.java]: We've made a new commit changing this to final

**Comment by magnus-madsen** [comment]
Nice! I will merge it soon!

**Pull request merged**